### PR TITLE
Render Images from Data Sources Test Fix

### PR DESCRIPTION
**Pull request by M1kep**
Updates the test for completion on Render Images from Data Sources to be somewhat more strict.

And this has been tested on a local version of FCC.

### CODE REVIEW

**Comment by M1kep** [comment]
closes #7110

**Comment by ltegman** [comment]
Thanks for the PR! This test is moving in the right direction but it's a bit too specific and some of it is unnecessarily complicated. Some suggestions:
+ All your explicit quantifiers can either be removed (`{1}` is implied unless otherwise specified) or replaced with the `?` shorthand.
+ Anywhere that your regex checks for a single or double quote needs to be able to accept either.
+ Anywhere you check for spaces needs to accept any whitespace and you need to make sure you allow whitespace in every place where it would be valid.
+ You can't assume the user will use `+=` instead of `__ = __ + __`.
+ All your capture groups can be removed and simply be character literals.
+ You don't need to do things like `(i|I)` if you use the `i` flag at the end of your regex.

**Comment by M1kep** [comment]
@ltegman 
Updated RegEx
```regex
/html *(\+=) *\" *<(i|I)(m|M)(g|G) *(s|S)(r|R)(c|C) *= *'? *" *\+ *val.imageLink *\+ *" *'? *> *" *;/g
```
>All your explicit quantifiers can either be removed ({1} is implied unless otherwise specified) or replaced with the ? shorthand.
~~Can you give specific place where this issues arises in the aforementioned(I think that's the right use of this word...) RegEx~~
I think I've fixed this is the RegEx at the top of this comment.

>Anywhere that your regex checks for a single or double quote needs to be able to accept either.
The main issue I found trying to do this was that it would require checking if the begining quotes where double or single cause mixing them incorrectly won't work.

>Anywhere you check for spaces needs to accept any white space and you need to make sure you allow white space in every place where it would be valid.
I've updated the RegEx to work with something like the following(In concern to white space.)
```html
html      +=    "   <img   src =   '    "   +       val.imageLink   +   "   '   >       "    ;
```
>You can't assume the user will use += instead of __ = __ + __
Uhm..... Ok, will have to look more into this piece.

>You don't need to do things like (i|I) if you use the i flag at the end of your regex.
The issue I ran into with the 'i' flag was that, "val.imageLink" needs to be case-sensitive or else it won't actually render the images.

**Comment by raisedadead** [comment]
@M1kep Any updates on this?

**Comment by M1kep** [comment]
@raisedadead @ltegman never responded to my inquiries/issues in my quoted response.

**Comment by geoff616** [comment]
@M1kep maybe something like this would work:
``` assert(code.match(/val.imageLink/g) && assert(code.match(html\s*(?:\+=|=\s*html\s*\+\s*)\s*(?:"|')\s*<\s*img\s+src\s*=\s*(?:"\s*'|'\s*")\s*\+\s*val\.imageLink\s*\+\s*(?:"\s*'|'\s*")\s*>\s*(?:"|')\s*;\s*/gi```where there is one case sensitive check for the variable `val.imageLink` and another case intensive regex to validate the HTML?

**Comment by BerkeleyTrue** [comment]
Can we get one of the @FreeCodeCamp/issue-moderators to QA

File: seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json [hunk 1]
```diff
@@ -340,7 +340,7 @@
         "</div>"
       ],
       "tests": [
-        "assert(code.match(/val.imageLink/gi), 'message: You should have used the <code>imageLink</code> property to display the images.');"
+        "assert(code.match(/(html) *(\\+=) *\\\"<{1}(i|I)(m|M)(g|G) *(s|S)(r|R)(c|C) *= *'{0,1}\" *\\+ *(val.imageLink) *\\+ *\" *'{0,1} *>\";/g), 'message: You should have used the <code>imageLink</code> property to display the images.');"
```
Comment (SaintPeter) [inline seed/challenges/01-front-end-development-certification/json-apis-and-ajax.json]: This is an overly complex regex.  
You should never need `{1}` because simply putting a character means one and only one of a character.

There is no need for all of the capture groups to get upper and lower case letter, instead just use case insensitive flag (`/ig`).

`{0,1}` can be replaced with `?` (which means zero or one of the preceding character).

Additionally, I see ` *` (Space star) which is probably more properly `\s*` to allow for tabs or even line-breaks.

I'm also not sure why there are capture groups around some internal strings, like `(val.imageLink)` - also the `.` in that string should be escaped.

**Comment by SaintPeter** [comment]
I would not accept this PR without a significant overhaul to the regex.  It may not be salvageable.

**Comment by raisedadead** [comment]
@M1kep Would you consider the comments and update this?

**Comment by raisedadead** [comment]
This PR not going to be accepted without changes as pointed in comments above.
Closing as stale as well, with no updates.

Feel free to open a fresh PR after addressing the comments.
Thanks!
